README.md
[![Build Status](https://travis-ci.org/resplab/cfmortality.svg?branch=master)](https://travis-ci.org/resplab/cfmortality)
[![CRAN Status](https://www.r-pkg.org/badges/version/cfmortality)](https://CRAN.R-project.org/package=cfmortality)
[![Project Status: Active – The project has reached a stable, usable state and is being actively developed.](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)

## cfmortality
Allows clinicians to predict 1- and 2- year risk of death (with  a threshold risk of death of >= 20% for the 1-year model) in cyctic fibrosis patients based on patients' overal health status described in [https://erj.ersjournals.com/content/54/3/1900224](https://erj.ersjournals.com/content/54/3/1900224).


### Installation
You can install the latest development version from GitHub:

```
install.package("remotes")
remotes::install_github("resplab/cfmortality")
```

### Mortality Prediction

To get a prediction for mortality rate for first and second year , you will need to pass in patient's risk factors. For example: 

```
predictcfmortality (age = 16, male = 0, fvc = 66.7, fev1 = 47.4, fev1LastYear = 80.5, bcepacia = 0, underweight = 0, nHosp = 0, pancreaticInsufficient = 1, CFRelatedDiabetes = 0, ageAtDiagnosis = 0.9)

```

The **predictcfmortality()** function returns 1- year and 2-year mortality rate of patients with cystic fibrosis with 20% cut-off for risk of death of the 1-year model.

### Cloud-based API Access
The [PRISM platform](http://prism.resp.core.ubc.ca) allows users to access CFMortality through the cloud. Fore more info please refer to the [PRISM model repository](http://resp.core.ubc.ca/ipress/prism).

#### R

In R, you can use package [`prism`](https://github.com/resplab/prism) to access the API.

```
library(prism)
connect_to_model('CFMortalityPrism', 123456, "cfmortality.cp.prism-ubc.linaralabs.com/ocpu/library/CFMortalityPrism/R/gateway/json")
input <- get_default_input()
model_run(input)
```

#### Linux

In Ubuntu, you can call the API with `curl`:

```
curl -X POST -H "Content-Type: application/json" -d '{"api_key":["123456"],"func":["prism_model_run"],"model_input":[{"age": 16,"male": 0,"age": 57,"fvc": 66.7,"fev1": 47.4,"fev1LastYear": 80.5,"bcepacia": 0,"underweight": 0,"nHosp": 0,"pancreaticInsufficient": 1,"CFRelatedDiabetes": 0,"ageAtDiagnosis": 0.9}]}' http://cfmortality.cp.prism-ubc.linaralabs.com//ocpu/library/CFMortalityPrism/R/gateway/json
```

#### Windows

In Windows PowerShell, you can use `curl` to access the API:

```
curl -Body '{"api_key":["123456"],"func":["prism_model_run"],"model_input":[{"age": 16,"male": 0,"age": 57,"fvc": 66.7,"fev1": 47.4,"fev1LastYear": 80.5,"bcepacia": 0,"underweight": 0,"nHosp": 0,"pancreaticInsufficient": 1,"CFRelatedDiabetes": 0,"ageAtDiagnosis": 0.9}]}' -Method POST -uri http://cfmortality.cp.prism-ubc.linaralabs.com//ocpu/library/CFMortalityPrism/R/gateway/json -Headers @{"Content-type"="application/json"}
```


### Citation

Please cite: 

```
Stanojevic, Sanja, et al. "Development and external validation of 1-and 2-year mortality prediction models in cystic fibrosis." European Respiratory Journal 54.3 (2019): 1900224.
```


NEWS.md
# cfmortality v0.3.0
* New release.

R/mpmcf.R
#' Predicting 1- and 2-year risk of death for patients with cystic fibrosis (CF)
#' @param age A number for age
#' @param male A binary variable with 0 for females and 1 for males
#' @param fvc  A number for percentage of FVC in current year
#' @param fev1  A number for percentage of fev11 in current year
#' @param fev1LastYear A number for percentage of fev11 in preceeding year
#' @param bcepacia A binary with 0 for no B. cepacia complex and 1 for B. cepacia complex
#' @param underweight A binary with 0 for not being underweight (BMI<18.5) and 1 otherwise
#' @param nHosp number of hospancreaicInsufficienttalizations in preceding year
#' @param pancreaicInsufficient A binary taking 1 for pancreaic insufficient status and 0 otherwise
#' @param CFRelatedDiabetes A binary variables for CF related diabetes
#' @param ageAtDiagnosis A number for age at CF diagnosis in years
#' @return The code
#' @examples
#' predictCFMortality (age = 16, male = 0, fvc = 66.7, fev1 = 47.4, fev1LastYear = 80.5, bcepacia = 0, underweight = 0, nHosp = 0, pancreaicInsufficient = 1, CFRelatedDiabetes = 0, ageAtDiagnosis = 0.9)
#' predictCFMortality(age = 40.4, male = 1, fvc = 25.7, fev1 = 19.2, fev1LastYear = 20, bcepacia = 1, underweight = 1, nHosp = 6, pancreaicInsufficient = 0, CFRelatedDiabetes = 0, ageAtDiagnosis = 27.2)
#' @source \url{https://erj.ersjournals.com/content/early/2019/05/08/13993003.00224-2019}
#' @export
predictCFMortality <- function(age, male, fvc, fev1, fev1LastYear, bcepacia, underweight, nHosp, pancreaicInsufficient, CFRelatedDiabetes, ageAtDiagnosis) {

  lny1 <-  5.702963 - 0.0162938*(male) + 0.7360137*log(fvc/100) + 0.7899955*log(fev1/100) - 0.7302478*(bcepacia) - 0.4588687*(underweight) - 0.0398486*(age) - 0.2818584*(nHosp)

  lnb1 <- 0.1146547 - 0.0792965*(nHosp) - 0.5616525*(log(fev1LastYear/100) -log(fev1/100)) + 0.2554754*(log(fev1/100)) + 0.6058589*(pancreaicInsufficient) + 0.2340407*(CFRelatedDiabetes) + 0.0079757*(ageAtDiagnosis)

  y1 <- exp(lny1)
  b1 <- exp(lnb1)

  lns1 <- (1/((-1)*b1))*(1/y1)^b1*(exp(b1)-1)

  s1 <- round(exp(lns1)*100, digits=2)

  lny2 <-  4.55962 + 0.3189947*(male) + 0.5809873*log(fvc/100) + 0.8404154*log(fev1/100) - 0.4187824*(underweight) - 0.9285728*(bcepacia)

  lnb2 <- 0.1863934 - 0.1263516*(nHosp) + 0.1858131*((log(fev1LastYear/100)) -log(fev1/100)) + 0.4353779*(log(fev1/100)) + 0.1927758*(pancreaicInsufficient) - 0.172767*(CFRelatedDiabetes) + 0.0012487*(ageAtDiagnosis)

  y2 <- exp(lny2)
  b2 <- exp(lnb2)

  lns2 <- (1/((-1)*b2))*(1/y2)^b2*(exp(b2)-1)

  s2 <- round(exp(lns2)*100, digits=2)

  osf <- round (s1*s2/100, digits = 2)


  p1 <- paste (" 1-year probability of survival is:",s1)
  p2 <- paste (" 2-year probability of survival is:", s2)
  p3 <- paste (" The overall probability of survival at 2 years is:",osf )

  cat(p1,p2,p3, sep = "\n")

}


R/CFMortality.R
#' Predicts 1- and 2- year Mortality Prediction Models in Cystic Fibrosis (CF)
#' 
#' @param age Patient age, in years
#' @param male A binary variable with 0 for females and 1 for males
#' @param fvc  FVC percent predicted in the current year (0-150)
#' @param fev1  FEV1 percent predicted in the current year (0-150)
#' @param fev1LastYear FEV1 percent predicted in the preceding year (0-150)
#' @param bcepacia A binary with 0 for no B. cepacia complex and 1 for B. cepacia complex
#' @param underweight A binary with 1 for underweight (BMI < 18.5 if age >= 19 or BMI percentile <= 12\% if age < 19)
#' @param nHosp An integer number of hospitalizations in preceding year
#' @param pancreaticInsufficient A binary taking 1 for pancreatic insufficient status and 0 otherwise
#' @param CFRelatedDiabetes A binary variable for CF related diabetes
#' @param ageAtDiagnosis A number for age at CF diagnosis in years
#' @return 1- and 2-year predicted mortality risk 
#' @examples
#' predictcfmortality (age = 16, male = 0, fvc = 66.7, fev1 = 47.4, fev1LastYear = 80.5, 
#'                     bcepacia = 0, underweight = 0, nHosp = 0, pancreaticInsufficient = 1, 
#'                     CFRelatedDiabetes = 0, ageAtDiagnosis = 0.9)
#' predictcfmortality (age = 40.4, male = 1, fvc = 25.7, fev1 = 19.2, fev1LastYear = 20, 
#'                     bcepacia = 1, underweight = 1, nHosp = 6, pancreaticInsufficient = 0, 
#'                     CFRelatedDiabetes = 0, ageAtDiagnosis = 27.2)
#' predictcfmortality (age = 44, male = 1, fvc = 72.95, fev1 = 55.5, fev1LastYear = 52.5, 
#'                     bcepacia = 0, underweight = 1, nHosp = 0, pancreaticInsufficient = 0, 
#'                     CFRelatedDiabetes = 0, ageAtDiagnosis = 29)
#' 
#' @source \url{https://erj.ersjournals.com/content/early/2019/05/08/13993003.00224-2019}
#' @export
predictcfmortality <- function (age, male, fvc, fev1, fev1LastYear, bcepacia, underweight, nHosp, pancreaticInsufficient, CFRelatedDiabetes, ageAtDiagnosis) {


  
  #1-year Mortality Prediction Model
  lny1 <-  5.702963 - 0.0162938*(male) + 0.7360137*log(fvc/100) + 0.7899955*log(fev1/100) - 0.7302478*(underweight) - 0.4588687*(bcepacia) - 0.0398486*(age) - 0.2818584*(nHosp)

  lnb1 <- 0.1146547 - 0.0792965*(nHosp) - 0.5616525*ifelse(fev1LastYear>fev1,1,0)*(log(fev1LastYear/100)-log(fev1/100)) + 0.2554754*(log(fev1/100)) + 0.6058589*(pancreaticInsufficient) + 0.2340407*(CFRelatedDiabetes) + 0.0079757*(ageAtDiagnosis)

  y1 <- exp(lny1)
  
  b1 <- exp(lnb1)

  lns1 <- (1/((-1)*b1))*((1/y1)^b1)*(exp(b1)-1)

  s1 <- round(exp(lns1)*100, digits=2)
  
  
  #2-year Mortality Prediction Model
  lny2 <-  4.55962 + 0.3189947*(male) + 0.5809873*log(fvc/100) + 0.8404154*log(fev1/100) - 0.4187824*(underweight) - 0.9285728*(bcepacia)

  lnb2 <- 0.1863934 - 0.1263516*(nHosp) + 0.1858131*ifelse(fev1LastYear>fev1,1,0)*(log(fev1LastYear/100)-log(fev1/100)) + 0.4353779*(log(fev1/100)) + 0.1927758*(pancreaticInsufficient) - 0.172767*(CFRelatedDiabetes) + 0.0012487*(ageAtDiagnosis)

  y2 <- exp(lny2)
  
  b2 <- exp(lnb2)

  lns2 <- (1/((-1)*b2))*((1/y2)^b2)*(exp(b2)-1)

  s2 <- round(exp(lns2)*100, digits=2)
  
  
  #Overal 2-year mortality
  os <- round (s1*s2/100, digits = 2)
  
  
  
  results <- list() 
  
  
  if (s1 >= 80.00) {
 
    results$first_year_survival <- s1
    results$second_year_survival <- s2
    results$overall_survival <- os
    return(results)
    
  } else {
    
    results$first_year_survival <- s1
    return(results)
    
  }  
}
